presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,274 @@
+            return Optional.empty();
+        }
+
+        JoinGraphNode joinGraph = (JoinGraphNode) node;
+        JoinNode joinNode = new JoinEnumerator(idAllocator, symbolAllocator, session, lookup).chooseJoinOrder(joinGraph);
+        return Optional.of(joinNode);
+    }
+
+    @VisibleForTesting
+    public class JoinEnumerator
+    {
+        private final Map<String, JoinNode> memo = new HashMap<>();
+        private final PlanNodeIdAllocator idAllocator;
+        private final SymbolAllocator symbolAllocator;
+        private final Session session;
+        private final Lookup lookup;
+
+        public JoinEnumerator(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session, Lookup lookup)
+        {
+            this.idAllocator = idAllocator;
+            this.symbolAllocator = symbolAllocator;
+            this.session = session;
+            this.lookup = lookup;
+        }
+
+        private JoinNode chooseJoinOrder(JoinGraphNode joinGraph)
+        {
+            if (memo.containsKey(joinGraph.getKey())) {
+                return memo.get(joinGraph.getKey());
+            }
+            else {
+                List<JoinNode> joinOrders = new ArrayList<>();
+                // TODO: don't consider partitions with unnecessary cross joins
+                for (Set<Integer> partitioning : generatePartitions(joinGraph.getSources().size())) {
+                    joinOrders.add(createJoinNodeTree(joinGraph, partitioning, idAllocator));
+                }
+                checkState(joinOrders.size() >= 1, "joinOrders cannot be empty");
+                JoinNode node = planNodeOrdering().min(joinOrders);
+                memo.put(joinGraph.getKey(), node);
+                return node;
+            }
+        }
+
+        private JoinNode createJoinNodeTree(JoinGraphNode joinGraph, Set<Integer> partitioning, PlanNodeIdAllocator idAllocator)
+        {
+            JoinNode placeholderJoin = createJoinAccordingToPartitioning(joinGraph, partitioning, idAllocator);
+            PlanNode left = placeholderJoin.getLeft();
+            PlanNode right = placeholderJoin.getRight();
+            if (left instanceof JoinGraphNode) {
+                left = chooseJoinOrder((JoinGraphNode) left);
+            }
+            if (right instanceof JoinGraphNode) {
+                right = chooseJoinOrder((JoinGraphNode) right);
+            }
+
+            JoinNode joinNode = (JoinNode) placeholderJoin.replaceChildren(ImmutableList.of(left, right));
+            return chooseJoinNodeProperties(joinNode);
+        }
+
+        private JoinNode chooseJoinNodeProperties(JoinNode joinNode)
+        {
+            List<JoinNode> possibleJoinNodes = new ArrayList<>();
+            String joinDistributionType = getJoinDistributionType(session);
+            if (joinDistributionType.equals(FeaturesConfig.JoinDistributionType.REPARTITIONED) || joinDistributionType.equals(FeaturesConfig.JoinDistributionType.AUTOMATIC)) {
```

Comment:
if you have enum with method then this like would just be:
```
if (joinDistributionType.isRepartitioned()) {
```
`isRepartitioned` would return true for both `AUTOMATIC` and `REPARTITIONED`